File: lib/data/database/data/instance/DataStore.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LastIRead.Data.Instance;
using LastIRead.windows.main.pages;
using LiteDB;

namespace LastIRead.data.database {
	public class DataStore : DatabaseCollection<IPersistentBookmark> {
		public DataStore(LiteDatabase database) : base(database, database.GetBookmarkCollection()) { }

		public IEnumerable<IUserBookmark> GetSelected(string filter, FilterData filterData) {
			var strippedFilter = StripString(filter);
			var result = Collection.FindAll();

			if (!string.IsNullOrEmpty(filter)) {
				result = result.Where(
					readable => Contains(readable.LocalizedTitle, strippedFilter) ||
					            Contains(readable.OriginalTitle, strippedFilter)
				);
			}

			if (filterData.Hide.HasFlag(Filter.Reading)) {
				result = result.Where(
					readable => !readable.Abandoned && (readable.Ongoing || readable.Progress < readable.MaxProgress)
				);
			}

			if (filterData.Hide.HasFlag(Filter.Abandoned)) {
				result = result.Where(readable => !readable.Abandoned);
			}

			if (filterData.Hide.HasFlag(Filter.Ended)) {
				result = result.Where(readable => readable.Ongoing);
			}

			if (filterData.Hide.HasFlag(Filter.Finished)) {
				result = result.Where(readable => readable.Ongoing || readable.Progress < readable.MaxProgress);
			}

			if (filterData.Hide.HasFlag(Filter.Ongoing)) {
				result = result.Where(readable => !readable.Ongoing);
			}

			return result
			       .Select(x => new WrapperUserBookmark(x))
			       .OrderBy(x => x.Title)
			       .ToArray();
		}

		private static bool Contains(string? title, string filter) {
			if (title == null) return false;

			var strippedTitle = StripString(title);
			return strippedTitle.Contains(filter, StringComparison.InvariantCultureIgnoreCase);
		}

		private static string StripString(string text) {
			var selectedCharacters = text
				.Where(
					character =>
						!char.IsPunctuation(character) &&
						!char.IsWhiteSpace(character) &&
						!char.IsSeparator(character)
				);

			return string.Concat(selectedCharacters);
		}
	}
}

File: lib/data/database/data/instance/DataStore.dart
import '../../../abstract/IPersistentBookmark.dart';
import '../abstract/DatabaseCollection.dart';
import '../../../runtime/FilterData.dart';
import '../../../extension/UserBookmark.dart';

class DataStore extends DatabaseCollection<IPersistentBookmark> {
  Future<Iterable<IPersistentBookmark>> GetSelected(
      String filter, FilterData filterData) async {
    filter = filter.toLowerCase();
    var strippedFilter = _stripString(filter);
    var result = await GetAll();

    if (filter.isNotEmpty) {
      result = result.where((readable) =>
          _contains(readable.localizedTitle, strippedFilter) ||
          _contains(readable.originalTitle, strippedFilter));
    }

    if (filterData.reading) {
      result = result.where((readable) =>
          !readable.abandoned &&
          (readable.ongoing || readable.progress < readable.maxProgress));
    }

    if (filterData.abandoned) {
      result = result.where((readable) => !readable.abandoned);
    }

    if (filterData.ended) {
      result = result.where((readable) => readable.ongoing);
    }

    if (filterData.finished) {
      result = result.where((readable) =>
          readable.ongoing || readable.progress < readable.maxProgress);
    }

    if (filterData.ongoing) {
      result = result.where((readable) => !readable.ongoing);
    }

    var resultList = result.toList();
    resultList.sort((a, b) => a.title.compareTo(b.title));
    return resultList;
  }

  static bool _contains(String title, String filter) {
    if (title == null) return false;

    var strippedTitle = _stripString(title);
    return strippedTitle.contains(filter);
  }

  static String _stripString(String text) {
    return text.replaceAll(RegExp('[\\p{IsPunctuation} ]'), text);
  }
}
